File: Week 1 - C/credit.c
// Learning C with CS50x
// Credit
// https://cs50.harvard.edu/x/2024/psets/1/credit/


// This program checks if the credit card is valid using Luhn’s CheckSum Algorithm and says to which credit card network the card belongs


#include <stdio.h>
#include <stdbool.h>

int last_digit(long long num);
int product_sum(int num);
bool checkSum(long long num);
void printCardName(long long num);

int main(void)
{
  printf("Give me the credit card number: ");
  long long num;
  scanf("%lld", &num);
  bool isValid = checkSum(num);
  if (isValid)
  {
    printCardName(num);
  }
  else
    printf("INVALID\n");
}

int last_digit(long long num) // returns that last digit of a number
{
  return (num % 10);
}

int product_sum(int num) // returns the sum of the product of a number. exemple : gets:6. 6*2=12. 1+2=3. return:3
{
  num *= 2;
  return num % 10 + num / 10;
}

bool checkSum(long long num) // check if the credit card is valid using Luhn’s CheckSum Algorithm
{
  int oddSum = 0;  // a sum of all the numbers that we haven't multiplied
  int evenSum = 0; // a sum off all the numbers that we multiplied by 2
  while (num > 0)
  {
    oddSum = oddSum + last_digit(num);
    num = num / 10;
    evenSum = evenSum + product_sum(last_digit(num));
    num = num / 10;
  }
  return last_digit(oddSum + evenSum) == 0;
}

int cardLength(long long num)
{
  int length = 0;
  while (num > 0)
  {
    num = num / 10;
    length++;
  }
  return length;
}

void printCardName(long long num)
{
  int length = cardLength(num);
  do
  {
    num = num / 10; // leave the 2 first digits
  } while (num > 99);

  if (length == 15) // check if it is AMEX
  {
    if (num == 34 || num == 37)
    {
      printf("AMEX\n");
      return;
    }
    else
    {
      printf("INVALID\n");
      return;
    }
  }

  if (length == 13) // check if it is VISA
  {
    if (num / 10 == 4)
    {
      printf("VISA\n");
      return;
    }
    else
    {
      printf("INVALID\n");
      return;
    }
  }

  if (length == 16) // check if it is VISA or MASTERCARD
  {
    if (num == 51 || num == 52 || num == 53 || num == 54 || num == 55)
    {
      printf("MASTERCARD\n");
      return;
    }
    else if (num / 10 == 4)
    {
      printf("VISA\n");
      return;
    }
  }
  else
  {
    printf("INVALID\n");
    return;
  }
}

File: Week 6 - Python/credit.py
# Learning Python with CS50x
# Credit
# https://cs50.harvard.edu/x/2024/psets/6/credit/


import re


def main():
    credit_card_number = int(input("Number: "))
    if not is_valid_credit_card_number(credit_card_number):
        print("INVALID")

    else:
        print(credit_card_company_name(credit_card_number))


def is_valid_credit_card_number(credit_card_number: int):
    # Get the length of the credit card number
    card_number_length = len(str(credit_card_number))

    valid_card_lengths = [13, 15, 16]

    # Check that the credit card number is exactly 16 characters
    if card_number_length not in valid_card_lengths:
        return False

    credit_card_number = str(credit_card_number)

    nDigits = len(credit_card_number)
    nSum = 0
    isSecond = False

    for i in range(nDigits - 1, -1, -1):
        d = ord(credit_card_number[i]) - ord("0")

        if isSecond == True:
            d = d * 2

        # We add two digits to handle
        # cases that make two digits after
        # doubling
        nSum += d // 10
        nSum += d % 10

        isSecond = not isSecond

    if nSum % 10 == 0:
        return True
    else:
        return False


def credit_card_company_name(credit_card_number: int) -> str:
    # Get the length of the credit card number
    card_number_length = len(str(credit_card_number))
    # Get the first two digits of the credit card number
    first_two_digits = str(credit_card_number)[0:2]

    # Check if the credit card number is a valid Visa card
    if (card_number_length == 13 or card_number_length == 16) and (
        first_two_digits[:1] == "4"
    ):
        return "VISA"

    # Check if the credit card number is a valid Mastercard card
    if (card_number_length == 16) and (
        first_two_digits == "51"
        or first_two_digits == "52"
        or first_two_digits == "53"
        or first_two_digits == "54"
        or first_two_digits == "55"
    ):
        return "MASTERCARD"

    # Check if the credit card number is a valid American Express card
    if (card_number_length == 15) and (
        first_two_digits == "34" or first_two_digits == "37"
    ):
        return "AMEX"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
